### PR TITLE
Implement low-hanging performance improvements from feedback investigation

### PR DESCRIPTION
**Pull request by sharwell**
Total allocations in the trace were 16GB.

Each of the first four commits saves allocations of >600MB.

Each of the last three commits saves allocations of >200MB.

Total improvement >20%.

Closes [AB#1224834](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1224834)

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/ConcurrentDictionaryExtensions.cs [hunk 1]
```diff
@@ -19,8 +21,22 @@ internal static class ConcurrentDictionaryExtensions
...
+        public static TValue GetOrAdd<TKey, TArg, TValue>(this ConcurrentDictionary<TKey, TValue> dictionary, TKey key, Func<TKey, TArg, TValue> valueFactory, TArg factoryArgument)
+            where TKey : notnull
+        {
+#if NETCOREAPP
+            return dictionary.GetOrAdd(key, valueFactory, factoryArgument);
+#else
+            if (dictionary.TryGetValue(key, out var value))
+                return value;
+
+            using var _ = PooledDelegates.GetPooledFunction(valueFactory, factoryArgument, out var boundFunction);
```
Comment (333fred) [inline src/Compilers/Core/Portable/InternalUtilities/ConcurrentDictionaryExtensions.cs:37]: Where is the `boundFunction` actually used?

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/ConcurrentDictionaryExtensions.cs:37]: I was afraid I might do this, hence adding a test for it 🤣

Comment (sharwell) [reply src/Compilers/Core/Portable/InternalUtilities/ConcurrentDictionaryExtensions.cs:37]: Amended the first commit to correct this line.

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder_GlobalSuppressions.cs [hunk 2]
```diff
@@ -254,42 +238,114 @@ static bool SupportsGlobalSuppression(ISymbol symbol)
...
+        /// <summary>
+        /// Validate and split a documentation comment ID into a prefix and complete symbol ID. For the
+        /// <paramref name="docCommentId"/> <c>~M:C.X(System.String)</c>, the <paramref name="prefix"/> would be
+        /// <c>~M:</c> and <paramref name="id"/> would be <c>C.X(System.String)</c>.
+        /// </summary>
+        [PerformanceSensitive("https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1224834", Constraint = "Avoid Regex splitting due to high allocation costs.")]
+        private static bool ValidateAndSplitDocumentationCommentId(
+            [NotNullWhen(true)] string? docCommentId,
+            out ReadOnlyMemory<char> prefix,
+            out ReadOnlyMemory<char> id)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder_GlobalSuppressions.cs:256]: curious why ROM instead of ROS?

**Review by CyrusNajmabadi — APPROVED** [review]
awesome!

**Review by 333fred — APPROVED** [review]
Compiler-side changes LGTM

File: src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder_GlobalSuppressions.cs (hunk 2 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder_GlobalSuppressions.cs:256]: `ReadOnlySpan<char>` cannot be assigned to a local variable in an `async` method (which the caller is).

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/ConcurrentDictionaryExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        public void TestGetOrAdd()
+        {
+            var first = new object();
+            var second = new object();
+
+            var dictionary = new ConcurrentDictionary<int, object>();
+            Assert.Same(first, dictionary.GetOrAdd(0, static (key, arg) => arg, first));
+            Assert.Same(first, dictionary[0]);
+            Assert.Same(first, dictionary.GetOrAdd(0, static (key, arg) => arg, second));
+            Assert.Same(first, dictionary[0]);
```
Comment (AlekseyTs) [inline src/Compilers/Core/CodeAnalysisTest/InternalUtilities/ConcurrentDictionaryExtensionsTests.cs:34]: >dictionary[0] [](start = 31, length = 13)

Consider asserting that count hasn't changed.

**Review by AlekseyTs — APPROVED** [review]
Changes under Compilers LGTM (iteration 7)

File: src/Compilers/Core/CodeAnalysisTest/InternalUtilities/ConcurrentDictionaryExtensionsTests.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/Core/CodeAnalysisTest/InternalUtilities/ConcurrentDictionaryExtensionsTests.cs:35]: ➡️ Amended first commit to include these assertions

**Pull request merged**